### PR TITLE
Unmarshal the config object

### PR DESCRIPTION
**Pull request by kislaykishore**
### Description
Construct the config object which is derived from the commandline
args, config file and the defaults.
### Link to the issue in case of a bug fix.
NA

### Testing details
1. Manual - NA
2. Unit tests - NA
3. Integration tests - NA

### CODE REVIEW

File: cmd/root.go [hunk 1]
```diff
@@ -42,5 +50,26 @@ func Execute() {
 }
 
 func init() {
-	// TODO: add flags and bind them with viper here.
+	cobra.OnInitialize(initConfig)
```
[the remark this answers is not in the document]
Comment (kislaykishore) [reply cmd/root.go:58]: initConfig is supposed to be executed by cobra. It's invoked when the command's Execute method is called and should not be called explicitly. There is some discussion [here](https://github.com/spf13/cobra/issues/1176#issuecomment-668158340).

File: cmd/root.go [hunk 2]
```diff
@@ -42,5 +50,26 @@ func Execute() {
...
+	if err != nil || cfgFile == "" {
+		return
+	}
+	viper.SetConfigFile(cfgFile)
+	viper.SetConfigType("yaml")
+	err = viper.ReadInConfig()
+	if err != nil {
+		return
+	}
+	err = viper.Unmarshal(&configObj, viper.DecodeHook(cfg.DecodeHook()), func(decoderConfig *mapstructure.DecoderConfig) {
```
[the remark this answers is not in the document]
Comment (kislaykishore) [reply cmd/root.go]: We are propagating this error which will result ultimately in logPanic logging this. Do you think we still need to log it?

File: cmd/root.go [hunk 3]
```diff
@@ -42,5 +50,26 @@ func Execute() {
...
+
+func initConfig() {
+	if err != nil || cfgFile == "" {
+		return
+	}
+	viper.SetConfigFile(cfgFile)
+	viper.SetConfigType("yaml")
+	err = viper.ReadInConfig()
+	if err != nil {
+		return
```
[the remark this answers is not in the document]
Comment (ashmeenkaur) [reply cmd/root.go:75]: won't this silently ignore the log file?

[the remark this answers is not in the document]
Comment (kislaykishore) [reply cmd/root.go]: Added logger.Fatal to exit in error cases.

File: cmd/root.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (kislaykishore) [reply cmd/root.go:51]: Resolving this as discussed offline.

File: cmd/root.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (kislaykishore) [reply cmd/root.go]: We are now exiting whenever an error is encountered in config.

**Review by ashmeenkaur — APPROVED** [review]

**Pull request merged**